src/main/java/io/lumigo/core/instrumentation/impl/ApacheKafkaProducerInstrumentation.java

```diff
@@ -0,0 +1,119 @@
+                                                .and(
+                                                        takesArgument(
+                                                                1,
+                                                                named(
+                                                                        "org.apache.kafka.clients.producer.Callback")))),
+                        ApacheKafkaProducerAdvice.class.getName());
+    }
+
+    public static class ApacheKafkaProducerAdvice {
+        public static final SpansContainer spansContainer = SpansContainer.getInstance();
+        public static final LRUCache<Integer, Boolean> handled = new LRUCache<>(1000);
+
+        @Advice.OnMethodEnter
+        public static <K, V> void methodEnter(
+                @Advice.FieldValue("metadata") ProducerMetadata metadata,
+                @Advice.FieldValue("keySerializer") Serializer<K> keySerializer,
+                @Advice.FieldValue("valueSerializer") Serializer<V> valueSerializer,
+                @Advice.Argument(value = 0, readOnly = false) ProducerRecord<K, V> record,
+                @Advice.Argument(value = 1, readOnly = false) Callback callback) {
+            try {
+                callback =
+                        new KafkaProducerCallback<>(
+                                callback,
+                                keySerializer,
+                                valueSerializer,
+                                metadata,
+                                record,
+                                System.currentTimeMillis());
+
+                // Try to inject correlation id to the kafka record headers
+                record.headers()
+                        .add(
+                                KafkaSpan.LUMIGO_MESSAGE_ID_KEY,
+                                UUID.randomUUID()
+                                        .toString()
+                                        .substring(0, 10)
+                                        .getBytes(StandardCharsets.UTF_8));
+            } catch (Exception e) {
+                Logger.error(e);
+            }
+        }
+
+        @AllArgsConstructor
+        public static class KafkaProducerCallback<K, V> implements Callback {
+            private final Callback callback;
+            private final Serializer<K> keySerializer;
+            private final Serializer<V> valueSerializer;
+            private final ProducerMetadata producerMetadata;
+            private final ProducerRecord<K, V> record;
+            private final long startTime;
+
+            @Override
+            public void onCompletion(RecordMetadata recordMetadata, Exception exception) {
+                try {
+                    if (callback != null) {
+                        callback.onCompletion(recordMetadata, exception);
+                    }
+                    Logger.info("Handling kafka request {}", record.hashCode());
+                    spansContainer.addKafkaProduceSpan(
```

Comment:
i'm not sure if it's good to have the spancontainer call the factory. maybe it's best to first create the span in the instrumentation and then add to the span container